CI: Run miri on each pull request

roypat opened the pull request:
### Summary of the PR

This PR adds miri to the CI for vm-memory, to ensure that we do not regress on recently fixed memory unsoundnesses (#228, #217). It fixes some more unsoundness issues in doctests, as well as some alignments issues that arise from miri's mock allocator.

Note that this blocked on https://github.com/rust-vmm/rust-vmm-ci/pull/126. 

### Requirements

Before submitting your PR, please make sure you addressed the following
requirements:

- [x] All commits in this PR are signed (with `git commit -s`), and the commit
  message has max 60 characters for the summary and max 75 characters for each
  description line.
- [x] All added/changed functionality has a corresponding unit/integration
  test.
- [x] All added/changed public-facing functionality has entries in the "Upcoming 
  Release" section of CHANGELOG.md (if no such section exists, please create one).
- [x] Any newly added `unsafe` code is properly documented.

andreeaflorescu commented:
@roypat it looks like it in the CI, we will need to publish a new container version that also has `rust-src`:
```
$ docker run -t -i --rm --init --volume /var/lib/buildkite-agent/builds/ip-172-31-29-42-4/rust-vmm/vm-memory-ci:/workdir --workdir /workdir --label com.buildkite.job-id=018842d7-a20d-4948-8ef8-41a4dc30eac1 rustvmm/dev:v23 /bin/sh -e -c RUST_BACKTRACE=1\ MIRIFLAGS=\'-Zmiri-disable-isolation\ -Zmiri-panic-on-unsupported\ -Zmiri-backtrace=full\'\ cargo\ +nightly\ miri\ test\ --features\ backend-mmap
--
  | I will run `"rustup" "component" "add" "rust-src"` to install the `rust-src` component for the selected toolchain. Proceed? [Y/n]
```

Can you test it with a local container build to make sure we have all the required tools?

roypat commented:
> @roypat it looks like it in the CI, we will need to publish a new container version that also has `rust-src`:
> 
> ```
> $ docker run -t -i --rm --init --volume /var/lib/buildkite-agent/builds/ip-172-31-29-42-4/rust-vmm/vm-memory-ci:/workdir --workdir /workdir --label com.buildkite.job-id=018842d7-a20d-4948-8ef8-41a4dc30eac1 rustvmm/dev:v23 /bin/sh -e -c RUST_BACKTRACE=1\ MIRIFLAGS=\'-Zmiri-disable-isolation\ -Zmiri-panic-on-unsupported\ -Zmiri-backtrace=full\'\ cargo\ +nightly\ miri\ test\ --features\ backend-mmap
> --
>   | I will run `"rustup" "component" "add" "rust-src"` to install the `rust-src` component for the selected toolchain. Proceed? [Y/n]
> ```
> 
> Can you test it with a local container build to make sure we have all the required tools?

Yep, I noticed that too, sorry >.> 

I added a temporary commit (f7805ad) that installs "rust-src" before executing miri and that makes the CI pass, so I went back to rust-vmm-container and added rust-src to it (https://github.com/rust-vmm/rust-vmm-container/pull/84)

andreeaflorescu commented:
> Yep, I noticed that too, sorry >.>
> 
> I added a temporary commit ([f7805ad](https://github.com/rust-vmm/vm-memory/commit/f7805addb5e56548863767471f0475030529711f)) that installs "rust-src" before executing miri and that makes the CI pass, so I went back to rust-vmm-container and added rust-src to it ([rust-vmm/rust-vmm-container#84](https://github.com/rust-vmm/rust-vmm-container/pull/84))

I just saw that afterwards, I approved the PR. But I see that it's currently stuck in testing.

roypat commented:
@andreeaflorescu every worked this time :partying_face:

roypat commented:
> Same issue with miri missing from the container image, this time on aarch64 though.

huh, that's bizarre... it was in there earlier :thinking: did we build another container since then?

roypat commented:
It seems like its running version 18 of the container, but it should be running version 24 (the newest)

`docker run -t -i --rm --init --volume /var/lib/buildkite-agent/builds/ip-172-31-72-62-1/rust-vmm/vm-memory-ci:/workdir --workdir /workdir --label com.buildkite.job-id=01885865-d0c8-4c71-930f-98389dd9c872 rustvmm/dev:v18 /bin/sh -e -c RUST_BACKTRACE=1\ MIRIFLAGS=\'-Zmiri-disable-isolation\ -Zmiri-panic-on-unsupported\ -Zmiri-backtrace=full\'\ cargo\ +nightly\ miri\ test\ --features\ backend-mmap`

roypat commented:
Ah, it seems like https://github.com/rust-vmm/vm-memory/pull/223 accidentally downgraded the container version - I'll fix it up
